src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -472,16 +522,32 @@ structure is_cycle {u : V} (p : G.walk u u)
 lemma is_trail_def {u v : V} (p : G.walk u v) : p.is_trail ↔ p.edges.nodup :=
 ⟨is_trail.edges_nodup, λ h, ⟨h⟩⟩
 
+@[simp] lemma is_trail_copy {u v u' v'} (p : G.walk u v) (hu : u = u') (hv : v = v') :
+  (p.copy hu hv).is_trail ↔ p.is_trail := by { subst_vars, refl }
+
 lemma is_path.mk' {u v : V} {p : G.walk u v} (h : p.support.nodup) : is_path p :=
 ⟨⟨edges_nodup_of_support_nodup h⟩, h⟩
 
 lemma is_path_def {u v : V} (p : G.walk u v) : p.is_path ↔ p.support.nodup :=
 ⟨is_path.support_nodup, is_path.mk'⟩
 
+@[simp] lemma is_path_copy {u v u' v'} (p : G.walk u v) (hu : u = u') (hv : v = v') :
+  (p.copy hu hv).is_path ↔ p.is_path := by { subst_vars, refl }
+
+lemma is_circuit_def {u : V} (p : G.walk u u) :
+  p.is_circuit ↔ is_trail p ∧ p ≠ nil :=
+iff.intro (λ h, ⟨h.1, h.2⟩) (λ h, ⟨h.1, h.2⟩)
+
+@[simp] lemma is_circuit_copy {u u'} (p : G.walk u u) (hu : u = u') :
```

Comment:
I like it how it is, it means we can be sure we aren't missing copy lemmas for any operation on graphs.

Reply:
Sure, but it also means it's harder to find lemmas about any other graph operations through the file, because there's `copy` lemmas scattered throughout. It's a question of do we want the docs to be easy to read, or do we want to layout the file to make sure lemmas aren't omitted? I think readability is more important here